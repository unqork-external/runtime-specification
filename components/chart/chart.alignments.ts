export enum ChartAlignments {
  LEFT = 'left',
  CENTER = 'center',
  RIGHT = 'right',
}
