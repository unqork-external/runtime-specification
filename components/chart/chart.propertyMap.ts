import { Description, Required } from '@tsed/schema'

export class ChartPropertyMap {
  @Required()
  @Description('Label or value for the X axis.')
  x: string

  @Required()
  @Description('Value for the Y axis.')
  y: string
}
