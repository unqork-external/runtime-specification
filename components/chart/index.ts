export * from './chart.alignments'
export * from './chart.definition'
export * from './chart.field'
export * from './chart.legend'
export * from './chart.propertyMap'
export * from './chart.refreshTarget'
export * from './chart.series'
export * from './chart.styling'
export * from './chart.targets.enum'
export * from './chart.types'
export * from './chart.verticalAlignments'
