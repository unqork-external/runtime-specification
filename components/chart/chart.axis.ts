import { Description, Optional, Property } from '@tsed/schema'

import { ChartText } from './chart.text'

export class ChartAxis {
  @Optional()
  @Description('A soft maximum for the axis.')
  softMax?: number

  @Optional()
  @Description('A soft minimum for the axis.')
  softMin?: number

  @Property(ChartText)
  @Optional()
  @Description("The chart's main title.")
  title?: ChartText
}
