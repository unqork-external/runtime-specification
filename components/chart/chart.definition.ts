import { CollectionOf, Const, DiscriminatorValue, Optional, Property, Required } from '@tsed/schema'

import { ChartField } from './chart.field'
import { ChartSeries } from './chart.series'
import { ChartStyling } from './chart.styling'
import { ChartTargets } from './chart.targets.enum'
import { ViewTargets } from '../../../decorators'
import { BaseComponentDefinition } from '../../base-component-interface'
import { Display } from '../../component-composition'
import type { SignalTargets } from '../../signals'

@DiscriminatorValue('vgChart')
@ViewTargets(ChartTargets)
export class ChartComponentDefinition extends BaseComponentDefinition {
  @Const('vgChart')
  type: 'vgChart' = 'vgChart' as const

  @Required()
  @CollectionOf(ChartSeries)
  series: ChartSeries[]

  @Required()
  @Property(Display)
  display = new Display()

  @Required()
  @Property(ChartField)
  field = new ChartField()

  @Optional()
  declare styling?: ChartStyling

  @Optional()
  declare signals?: SignalTargets<ChartTargets>
}
