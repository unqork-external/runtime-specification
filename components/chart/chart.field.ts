import { Description, Optional, Property, Required } from '@tsed/schema'

import { ChartAxis } from './chart.axis'
import { ChartLegend } from './chart.legend'
import { ChartText } from './chart.text'

export class ChartField {
  @Optional()
  @Description('Accessible label that describes the chart for screen readers.')
  ariaLabel?: string

  @Optional()
  @Property(ChartLegend)
  @Description('Legend configuration.')
  legend = new ChartLegend()

  @Property(ChartText)
  @Optional()
  @Description("The chart's subtitle.")
  subtitle?: ChartText

  @Property(ChartText)
  @Optional()
  @Description("The chart's main title.")
  title?: ChartText

  @Property(ChartAxis)
  @Required()
  @Description('Configuration for the X axis.')
  xAxis: ChartAxis

  @Property(ChartAxis)
  @Required()
  @Description('Configuration for the Y axis.')
  yAxis: ChartAxis
}
