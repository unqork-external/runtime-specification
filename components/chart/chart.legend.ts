import { Default, Description, Enum, Optional } from '@tsed/schema'

import { ChartAlignments } from './chart.alignments'
import { ChartVerticalAlignments } from './chart.verticalAlignments'

export class ChartLegend {
  @Optional()
  @Enum(ChartAlignments)
  @Default(ChartAlignments.CENTER)
  @Description('The horizontal alignment of the legend box within the chart area.')
  alignment: ChartAlignments = ChartAlignments.CENTER

  @Default(false)
  @Description('Whether to hide the chart legend.')
  hidden: boolean = false

  @Optional()
  @Enum(ChartVerticalAlignments)
  @Default(ChartVerticalAlignments.BOTTOM)
  @Description('The vertical alignment of the legend box.')
  verticalAlignment: ChartVerticalAlignments = ChartVerticalAlignments.BOTTOM
}
