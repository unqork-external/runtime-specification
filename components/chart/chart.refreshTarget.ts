import { Default, Description, Optional, Required } from '@tsed/schema'

import { ChartPropertyMap } from './chart.propertyMap'

export class ChartRefreshTarget {
  @Required()
  @Description('Property ID of the referenced component where the data resides.')
  target: string

  @Optional()
  @Description('Value of the referenced component where the data resides.')
  @Default('value')
  property = 'value'

  @Required()
  @Description(
    'Property map indicating which properties in the data reference component value array map to x and y values.',
  )
  propertyMap: ChartPropertyMap
}
