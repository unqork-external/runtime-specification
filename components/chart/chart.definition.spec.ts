import { ChartComponentDefinition } from './chart.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('Vega Chart Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(ChartComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
  })
})
