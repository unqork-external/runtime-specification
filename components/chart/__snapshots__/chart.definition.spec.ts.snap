// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Vega Chart Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
  "series",
  "display",
  "field",
]
`;

exports[`Vega Chart Contract Tests should ensure the contract is clear 2`] = `
{
  "display": {
    "$ref": "#/definitions/Display",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "field": {
    "$ref": "#/definitions/ChartField",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "series": {
    "items": {
      "$ref": "#/definitions/ChartSeries",
    },
    "type": "array",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/ChartStyling",
  },
  "type": {
    "const": "vgChart",
    "description": "Name of the type of component.",
    "examples": [
      "vgChart",
    ],
    "minLength": 1,
    "type": "string",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "Current value of the component",
    "type": [
      "null",
      "integer",
      "number",
      "string",
      "boolean",
      "array",
      "object",
    ],
  },
}
`;
